test/spmd/tensor/test_common_rules.py

```diff
@@ -251,6 +266,47 @@ def test_pointwise_rules_suggestion(self):
         self.assertEqual(len(schema_suggestion.args_schema), 3)
         self.assertEqual(schema_suggestion.args_schema[2], -1)
 
+    @with_comms
+    def test_pointwise_multi_sharding_on_mesh_dim(self):
+        # 2d mesh einop sharding
+        mesh_shape = torch.arange(self.world_size).reshape(
+            self.world_size // 2, self.world_size // 2
+        )
+        mesh = DeviceMesh(self.device_type, mesh_shape)
+        # basic case to test implicit broadcasting shape alignment
+        mat1, mat2 = [-1, 0], [0]
+        mat1_spec = DTensorSpec.from_dim_map(
+            mesh, mat1, [], shape=torch.Size([20, 6])
+        )
+        mat2_spec = DTensorSpec.from_dim_map(
+            mesh, mat2, [], shape=torch.Size([6])
+        )
+        output_sharding = pointwise_rule(OpSchema((mat1_spec, mat2_spec), {}))
+        output_spec = output_sharding.output_spec
+        self.assertIsNotNone(output_spec)
+        self.assertEqual(output_spec.dim_map, [-1, 0])
+
+        # more advanced case that needs reshard one input to align sharding
+        mat1, mat2 = [0, -1, -1, 1], [0, -1, 1]
+        mat1_spec = DTensorSpec.from_dim_map(
+            mesh, mat1, [], shape=torch.Size([12, 1, 1, 8])
+        )
+        mat2_spec = DTensorSpec.from_dim_map(
+            mesh, mat2, [], shape=torch.Size([12, 4, 8])
+        )
+        output_sharding = pointwise_rule(OpSchema((mat1_spec, mat2_spec), {}))
+        output_spec = output_sharding.output_spec
+        self.assertIsNone(output_spec)
+        self.assertIsNotNone(output_sharding.schema_suggestions)
+
+        # ensure that the suggestion is to reshard the first
+        # arg by all_gather first tensor dim sharding
+        schema_suggestion = output_sharding.schema_suggestions[0]
+        self.assertEqual(
+            schema_suggestion.args_schema[0].dim_map, [-1, -1, -1, 1]
```

Comment:
Is this cheaper than [-1, 0, -1, 1] ?

Reply:
note that we can't shard to `[-1, 0, -1, 1]` because the size of this input is `torch.Size([12, 1, 1, 8])`

Reply:
got it , so it actually would do [-1, 0, -1, 1] if the dimension was not a singleton?

Reply:
yeah it should be as long as this reshard is cheaper in terms of comm volume, let me see if i can add a test case for this as well.

Reply:
I'll land this first to unblock your work and add more tests later.